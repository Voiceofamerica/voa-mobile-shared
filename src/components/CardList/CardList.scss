
@import '../../commonStyles/index.scss';

.cardList {
  flex: 1;
  display: flex;
  flex-direction: column;
  overflow-y: scroll;
  -webkit-overflow-scrolling: touch;
  overflow-x: hidden;

  .row {
    display: flex;
    flex-direction: column;

    &.fixed {
      height: calc(100vw * 9 / 16 + 11px);
    }
  }
}
