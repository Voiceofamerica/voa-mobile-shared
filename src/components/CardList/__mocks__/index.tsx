
import * as React from 'react'

function CardList (props) {
  return (
    React.createElement('card-list', { props }, props.children)
  )
}

namespace CardList {
  export const Static = (props) => (
    React.createElement('static-card-list', { props }, props.children)
  )
}

export default CardList
