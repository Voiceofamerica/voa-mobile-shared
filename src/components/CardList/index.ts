
export { default } from './CardList'
export { default as StaticCardList } from './StaticCardList'
export * from './CardList'
export * from './CardListTypes'
