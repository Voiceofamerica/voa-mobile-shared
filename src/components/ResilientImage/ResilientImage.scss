
@import '../../commonStyles/index.scss';

.resilientImage {
  position: relative;
  background: black;

  .imageContent {
    object-fit: cover;
    width: 100%;
    height: 100%;

    &.containImage {
      object-fit: contain;
      height: auto;
      max-height: 100%;
    }
  }

  .childrenArea {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    display: flex;
    justify-content: center;
    align-items: center;
  }

  .spinner {
    color: $voa-blue;
    height: 80%;
    max-height: 100px;
    margin-top: -1%;
  }
}
