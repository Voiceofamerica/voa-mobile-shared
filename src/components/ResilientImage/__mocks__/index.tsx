
import * as React from 'react'

export default (props) => (
  React.createElement('resilient-image', { props }, props.children)
)
