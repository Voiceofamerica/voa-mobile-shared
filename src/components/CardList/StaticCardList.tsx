
import * as React from 'react'

import Card from '../Card'

import { BaseProps } from './CardListTypes'
import { cardList, row, fixed } from './CardList.scss'

export interface Props extends BaseProps {
}

class StaticCardList extends React.Component<Props> {
  render () {
    const { items = [] } = this.props

    return (
      <div className={cardList}>
        {
          items.length > 0
          ? this.renderContent()
          : ''
        }
      </div>
    )
  }

  private renderContent = () => {
    const { items, onItemClick } = this.props

    return items.map(({ id, image, title, minorText }) => (
      <div key={id} className={`${fixed} ${row}`}>
        <Card
          onPress={() => onItemClick(id)}
          title={title}
          imageUrl={image && image.hero}
        />
      </div>
    ))
  }
}

export default StaticCardList
