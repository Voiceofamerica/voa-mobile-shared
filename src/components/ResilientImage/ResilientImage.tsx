
import * as React from 'react'

import { waitUntilOnline } from '../../helpers/resilience'

import Spinner from '../Spinner'

import { resilientImage, imageContent, childrenArea, containImage, spinner } from './ResilientImage.scss'

export interface Props {
  src: string | null
  showSpinner?: boolean
  className?: string
  alwaysShow?: boolean
  style?: React.CSSProperties
  defaultSrc?: string
  onLoadDone?: () => void
  onClick?: () => void
  contain?: boolean
}

export type ImageStatus = 'loading' | 'loaded' | 'error'

export interface State {
  shouldRender?: boolean
  imgUrl?: string
  imageStaus: ImageStatus
}

const noop = () => null
const IMAGE_FETCH_RETRY_RATE = 2000

class ReilientImage extends React.Component<Props, State> {
  state: State = {
    imageStaus: 'loading',
  }

  mounted: boolean = false

  readonly abortController = new AbortController()

  componentWillMount () {
    this.mounted = true
    const shouldRender = navigator.onLine || this.props.alwaysShow
    this.setState({
      shouldRender,
    })

    if (shouldRender) {
      this.tryFetchImage()
    }
  }

  componentWillUnmount () {
    this.mounted = false
    this.abortController.abort()
  }

  componentWillReceiveProps (nextProps: Props) {
    if (nextProps.src !== this.props.src) {
      this.setImageStatus('loading')

      const shouldRender = navigator.onLine || this.props.alwaysShow
      this.setState({
        shouldRender,
      })

      if (shouldRender) {
        this.tryFetchImage()
          .catch(err => {
            console.warn(`Something went wrong loading ${nextProps.src}`, err)
          })
      }
    }
  }

  tryFetchImage = async () => {
    const { src } = this.props

    if (!src || !this.mounted) {
      return
    }

    if (!navigator.onLine) {
      waitUntilOnline().then(() => this.mounted && this.tryFetchImage())
      return
    }

    const { onLoadDone = noop } = this.props

    try {
      const blobUrl = await fetch(src, {
          method: 'GET',
          signal: this.abortController.signal,
        })
        .then(res => res.blob())
        .then(blob => URL.createObjectURL(blob))

      this.setImageStatus('loaded', blobUrl)
      onLoadDone()

    } catch (err) {
      this.setImageStatus('error')
      setTimeout(this.tryFetchImage, IMAGE_FETCH_RETRY_RATE)
    }
  }

  setImageStatus = (imageStaus: ImageStatus, imgUrl?: string) => {
    if (!this.mounted) {
      return
    }

    this.setState({
      imageStaus,
      imgUrl,
    })
  }

  renderOffline = () => (
    <div style={{ width: 0, height: 0 }} />
  )

  renderContent = () => {
    const { imageStaus, imgUrl } = this.state
    const { defaultSrc = require('./imagedefault.gif'), contain } = this.props

    const isLoaded = imageStaus === 'loaded'
    const srcToUse = isLoaded && imgUrl
                   ? imgUrl
                   : defaultSrc

    const className = contain ? `${imageContent} ${containImage}` : imageContent

    return (<img src={srcToUse} className={className} />)
  }

  renderSpinner = () => {
    const { showSpinner } = this.props
    const { imageStaus } = this.state

    const isLoaded = imageStaus === 'loaded'

    if (isLoaded || !showSpinner) {
      return null
    }

    return (
      <div className={childrenArea}>
        <Spinner className={spinner} />
      </div>
    )
  }

  renderChildren = () => {
    const { children } = this.props

    return (
      <div className={childrenArea}>
        {children}
      </div>
    )
  }

  render () {
    const { shouldRender = navigator.onLine || this.props.alwaysShow } = this.state
    if (!shouldRender) {
      return this.renderOffline()
    } else {
      const {
        className = '',
        style,
        onClick = () => null,
      } = this.props

      return (
        <div className={`${resilientImage} ${className}`} style={style} onClick={ev => onClick()}>
          { this.renderContent() }
          { this.renderSpinner() }
          { this.props.children ? this.renderChildren() : null }
        </div>
      )
    }
  }
}

export default ReilientImage
