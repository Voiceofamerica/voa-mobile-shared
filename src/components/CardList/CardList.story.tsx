
import * as React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import StaticCardList from './StaticCardList'
import CardList from './CardList'
import { ListItem } from './CardListTypes'

const testItem: ListItem = {
  id: 1,
  image: {
    tiny: 'https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w125_r1.png',
    thumb: 'https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w250_r1.png',
    hero: 'https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w500_r1.png',
  },
  minorText: '15:08',
  title: 'President Trump: From North Korea to Iran, we confront dangerous governments',
}

const createList = (count: number) => {
  const output: ListItem[] = []
  for (let i = 0; i < count; i++) {
    output.push({
      ...testItem,
      id: i,
    })
  }
  return output
}

const list = createList(1000)
const shortList = createList(30)

const containerStyle: React.CSSProperties = {
  height: '100vh',
  display: 'flex',
  flexDirection: 'column',
}

storiesOf(CardList.name, module)
  .add('display', () => (
    <div style={containerStyle}>
      <style>{'body { margin: 0; }'}</style>
      <CardList
        items={list}
        onItemClick={action('onItemClick')}
      />
    </div>
  ))
  .add('rtl display', () => (
    <div>
      <style>{'body { margin: 0; }'}</style>
      <CardList
        items={list}
        onItemClick={action('onItemClick')}
        dir='rtl'
      />
    </div>
  ))
  .add('static height display', () => (
    <div>
      <style>{'body { margin: 0; }'}</style>
      <CardList
        items={list}
        onItemClick={action('onItemClick')}
        height={200}
      />
    </div>
  ))
  .add('contained height display', () => (
    <div style={{height: 300, display: 'flex'}}>
      <style>{'body { margin: 0; }'}</style>
      <CardList
        items={list}
        onItemClick={action('onItemClick')}
      />
    </div>
  ))

storiesOf(`${CardList.name}.Static`, module)
  .add('display', () => (
    <div>
      <style>{'body { margin: 0; }'}</style>
      <StaticCardList
        items={shortList}
        onItemClick={action('onItemClick')}
      />
    </div>
  ))
  .add('rtl display', () => (
    <div dir='rtl'>
      <style>{'body { margin: 0; }'}</style>
      <StaticCardList
        items={shortList}
        onItemClick={action('onItemClick')}
      />
    </div>
  ))
