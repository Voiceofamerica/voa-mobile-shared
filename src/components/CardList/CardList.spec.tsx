
import * as React from 'react'
import { create } from 'react-test-renderer'
import { shallow } from 'enzyme'

import { ListItem } from './CardListTypes'
import CardList, { Props, State } from './CardList'
import ResilientImage from '../ResilientImage'

jest.mock('../ResilientImage')
const testItem: ListItem = {
  id: 1,
  image: {
    hero: 'https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w650_r1.png',
  },
  minorText: '15:08',
  title: 'President Trump: From North Korea to Iran, we confront dangerous governments',
}

const createList = (count: number) => {
  const output: ListItem[] = []
  for (let i = 0; i < count; i++) {
    output.push({
      ...testItem,
      id: i,
    })
  }
  return output
}

const list = createList(1000)

describe(`<${CardList.name} />`, () => {
  describe('shapshots', () => {
    it('should render only the necessary items', () => {
      const element = create((
        <CardList items={list} onItemClick={jest.fn()} />
      ))

      expect(element).toMatchSnapshot()
    })
    it('should render the empty content if no items', () => {
      const element = create((
        <CardList items={[]} onItemClick={jest.fn()} emptyContent='I am empty' />
      ))

      expect(element).toMatchSnapshot()
    })
    it('should not throw if items is undefined', () => {
      const element = create((
        <CardList items={undefined} onItemClick={jest.fn()} />
      ))

      expect(element).toMatchSnapshot()
    })
  })
})
