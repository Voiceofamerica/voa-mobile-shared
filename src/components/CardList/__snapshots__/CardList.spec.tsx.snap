// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardList /> shapshots should not throw if items is undefined 1`] = `
<div
  className="cardList"
>
  
</div>
`;

exports[`<CardList /> shapshots should render only the necessary items 1`] = `
<div
  className="cardList"
>
  <react-virtualized-list
    height={618}
    rowCount={1000}
    rowHeight={NaN}
    rowRenderer={[Function]}
    width={1024}
  >
    <div
      className="row"
      dir="ltr"
      style={Object {}}
    />
    <div
      className="row"
      dir="ltr"
      style={Object {}}
    />
    <div
      className="row"
      dir="ltr"
      style={Object {}}
    />
    <div
      className="row"
      dir="ltr"
      style={Object {}}
    />
    <div
      className="row"
      dir="ltr"
      style={Object {}}
    />
  </react-virtualized-list>
</div>
`;

exports[`<CardList /> shapshots should render the empty content if no items 1`] = `
<div
  className="cardList"
>
  I am empty
</div>
`;
